docs: router request lifecycle (#6497)
apollographql/router: docs/source/routing/request-lifecycle.mdx

Author: The new top-level Router Request Lifecycle page

## The request lifecycle

Every client request made to a GraphOS Router goes through its **router request lifecycle**: a multi-stage pipeline of services that processes requests and returns responses.

<RequestLifecycleOverviewDiagram />

A client request is processed first by passing it between services along the lifecycle's **request path** to create and send its constituent subgraph requests. Then all the responses from the subgraph requests are gathered into a client response along the lifecycle's **response path**.

## Request path

In the request path, each request is processed by the request lifecycle services in the following order:

* The **Router service** receives the client request from the HTTP server and parses it into a GraphQL operation.
* The **Supergraph service** receives a GraphQL operation and calls the router's query planner to produce the query plan that most efficiently executes the operation.
* The **Execution service** executes a query plan by calling the necessary subgraph services to make subgraph requests
* Each subgraph has an associated **Subgraph service** that makes HTTP requests to the subgraph.

> Each service encapsulates and transforms the contents of a request into own context. The following diagram and its steps describe how an HTTP request is transformed and propagated through the request path:

Author: ```suggestion
Each service encapsulates and transforms the contents of a request into its own context. The following diagram and its steps describe how an HTTP request is transformed and propagated through the request path:
```

        end

        subgraph subgraphService["Subgraph Services"]
            subgraph service1["Subgraph Service A"]
            subgraphPlugins1[[Subgraph plugins]];
            end
            subgraph service2["Subgraph Service B"]
            subgraphPlugins2[[Subgraph plugins]];
            end            
        end
    end;
subgraphA[Subgraph A];
subgraphB[Subgraph B];

client --"1. HTTP request"--> httpServer;
httpServer --"2. <code>RouterRequest</code>"--> routerService;
routerService --"3. <code>SupergraphRequest</code>"--> supergraphService
supergraphService --"4. Query"--> queryPlanner;
queryPlanner --"5. Query plan"--> supergraphService;
supergraphService --"6. <code>ExecutionRequest</code>"--> executionService;

executionService --"7a. <code>SubgraphRequest</code>"--> service1;
executionService --"7b. <code>SubgraphRequest</code>"--> service2;

service1 --"8a. HTTP request"--> subgraphA;
service2 --"8b. HTTP request"--> subgraphB;
```

1. The router receives a client request at an HTTP server. 
2. The HTTP server transforms the HTTP request into a `RouterRequest` containing HTTP headers and the request body as a stream of byte arrays.

> 3. The router service receives the `RouterRequest`. It handles Automatic Persisted Queries (APQ), parses the GraphQL request from JSON, and calls the supergraph service with the resulting `SupergraphRequest`.

Reviewer: I didn't realize how "recent" this happened but the Router Service now also handles validation of the query against the schema - https://github.com/apollographql/router/pull/4551

> This removes GraphQL validation from the query planner, to use the Rust version instead. Validation has now moved to the query analysis layer...

Author: ```suggestion
3. The router service receives the `RouterRequest`. It handles Automatic Persisted Queries (APQ), parses the GraphQL request from JSON, validates the query against the schema, and calls the supergraph service with the resulting `SupergraphRequest`.
```

routerService --"13. <code>RouterResponse</code>"--> httpServer;
httpServer --"14. HTTP response" --> client
```
9. Each subgraph provides an HTTP response to the subgraph services.
10. Each subgraph service creates a `SubgraphResponse` containing the HTTP headers and a GraphQL response.
11. Once the execution service has received all subgraph responses, it formats the GraphQL responses—removing unneeded data and propagating nulls—before sending it back to the supergraph plugin as the `ExecutionResponse`.
12. The `SupergraphResponse` has the same content as the `ExecutionResponse`. It contains headers and a stream of GraphQL responses. That stream only contains one element for most queries—it can contain more if the query uses the `@defer` directive or subscriptions.
13. The router service receives the `SupergraphResponse` and serializes the GraphQL responses to JSON.
14. The HTTP server sends the JSON in an HTTP response to the client.

## Request and response nuances

Although the preceding diagrams showed the request and response paths separately and sequentially, in reality some requests and responses may happen simultaneously and repeatedly.

> For example, `SubgraphRequest`s can happen both in parallel _and_ in sequence: one subgraph's response may be necessary for another's `SubgraphRequest`. (The query planner decides which requests can happen in parallel vs. which need to happen in sequence). To match subgraph requests to responses in customizations, the router exposes a `subgraph_request_id` field that will hold the same value in paired requests and responses.

Reviewer: Might be helpful to provide a link to the query plan doc here? https://www.apollographql.com/docs/graphos/reference/federation/query-plans

Author: ```suggestion
For example, `SubgraphRequest`s can happen both in parallel _and_ in sequence: one subgraph's response may be necessary for another's `SubgraphRequest`. The [query planner](/graphos/reference/federation/query-plans) decides which requests can happen in parallel vs. which need to happen in sequence. 

To match subgraph requests to responses in customizations, the router exposes a `subgraph_request_id` field that will hold the same value in paired requests and responses.
```

    end
    subgraphA[Subgraph A];
    subgraphB[Subgraph B];

    service1 --"2. HTTP request"--> subgraphA;
    service2 --"6. HTTP request"--> subgraphB;
    subgraphA --"3. HTTP response"--> service1;
    subgraphB --"7. HTTP response"--> service2;
```

Additionally, some requests and responses may happen multiple times for the same operation. With subscriptions, for example, a subgraph sends a new `SubgraphResponse` whenever data is updated. Each response object travels through all the services in the response path and interacts with any customizations you've created.

### Request and Response buffering

The router expects to execute on a stream of data. In order to work correctly and provide high performance, the following expectations must be met:

* **Request Path**: No buffering before the end of the `router_service` processing step
* **Response Path**: No buffering

> In general, it's best to avoid buffering where possible. If necessary, it is ok to do so on the request path once the `router_service` step is complete.

This guidance applies if you are:
 - Modifying the router
 - Creating a native Rust plugin

>  - Creating a custom binary

Reviewer: What do you think about putting this first under the header so when reading you can make a decision about whether to continue or not based on whether it applies to you?

---
title: Router Request Lifecycle
subtitle: Understand how the router processes client requests
description: Understand how GraphQL client requests get processed by the request lifecycle pipeline of Apollo GraphOS Router.
---

import RequestLifecycleOverviewDiagram from '../../shared/diagrams/router-request-lifecycle-overview.mdx';

## The request lifecycle

Every client request made to a GraphOS Router goes through its **router request lifecycle**: a multi-stage pipeline of services that processes requests and returns responses.

<RequestLifecycleOverviewDiagram />

> A client request is processed first by passing it between services along the lifecycle's **request path** to create and send its constituent subgraph requests. Then all the responses from the subgraph requests are gathered into a client response along the lifecycle's **response path**.

Author: ```suggestion
The router processes a client request by first passing it between services along the lifecycle's **request path**. The request path breaks up a request and sends constituent sub-requests to subgraphs. Then along the lifecycle's **response path**, the router gathers all the responses from the subgraph requests into a client response.
```

> 

Reviewer 2: Before getting into the nitty-gritty in the following sections, is it worth calling out why users should _care_ about the lifecycle? (I.e. for instrumentation and customization)? I imagine most folks will come to this page because of these needs, but for the rare viewers who are browsing the top-level pages (About GraphOS Router, Configure Your Router, Request Lifecycle), it may be worth leading with this info.
